daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -255,18 +292,27 @@ class BucketingInputSource(
    */
   private def fillBucketsToIndex(goalBucketIndex: Long, bytesNeededInBucket: Long): Boolean = {
     var lastBucketIndex = buckets.length - 1
-    var hasMoreData = true
+
     var needsMoreData = goalBucketIndex > lastBucketIndex || (goalBucketIndex == lastBucketIndex && bytesNeededInBucket > bytesFilledInLastBucket)
 
     while (needsMoreData && hasMoreData) {
       // Try to fill the rest of this bucket, regardless of how many bytes are
       // actually needed
       val emptyBytesInLastBucket = bucketSize - bytesFilledInLastBucket
 
+      Assert.invariant(emptyBytesInLastBucket >= 0)
       // Try to read enough bytes to fill the rest of this bucket. Note that
       // the .read() function could hit EOF (returns -1) or could return
-      // anywhere from zero to emptyBytesInLastBucket bytes
-      val bytesRead = inputStream.read(buckets(lastBucketIndex).bytes, bytesFilledInLastBucket, emptyBytesInLastBucket)
+      // anywhere from 0 to emptyBytesInLastBucket bytes
```

Comment:
Should we update this comment to say 1 to emptyBytesInLastBucket byte? ``read()`` should never actually return 0.

Reply:
yes, done